### PR TITLE
[SPARK-41015][SQL][PROTOBUF] UnitTest null check for data generator 

### PR DESCRIPTION
**Pull request by SandishKumarHN**
### What changes were proposed in this pull request?
null check for data generator after type conversion.

### Why are the changes needed?
To fix a test failure.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
I have tested all the random numbers manually, current unit tests.

### CODE REVIEW

**Comment by SandishKumarHN** [comment]
@rangadi Because some random numbers do not convert to catalyst type, a null check for the data generator is required.

File: connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufCatalystDataConversionSuite.scala [hunk 1]
```diff
@@ -123,16 +123,21 @@ class ProtobufCatalystDataConversionSuite
...
       val generator = RandomDataGenerator.forType(dt, rand = rand).get
       var data = generator()
-      while (data.asInstanceOf[Row].get(0) == defaultValue) // Do not use default values, since
-        data = generator()                                  // from_protobuf() returns null in v3.
+      // Do not use default values, since from_protobuf() returns null in v3.
+      while (
+        data != null &&
+        (data.asInstanceOf[Row].get(0) == defaultValue ||
+          (data.asInstanceOf[Row].get(0).isInstanceOf[Array[Byte]] &&
+          data.asInstanceOf[Row].get(0).asInstanceOf[Array[Byte]].isEmpty)))
```
Comment (rangadi) [inline connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufCatalystDataConversionSuite.scala:139]: Isn't `ByteString.empty().toByteArray` at line 122 already empty value?

File: connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufCatalystDataConversionSuite.scala [hunk 2]
```diff
@@ -123,16 +123,21 @@ class ProtobufCatalystDataConversionSuite
...
       val (messageName, defaultValue) = catalystTypesToProtoMessages(dt.fields(0).dataType)
 
       val rand = new scala.util.Random(seed)
       val generator = RandomDataGenerator.forType(dt, rand = rand).get
       var data = generator()
-      while (data.asInstanceOf[Row].get(0) == defaultValue) // Do not use default values, since
-        data = generator()                                  // from_protobuf() returns null in v3.
+      // Do not use default values, since from_protobuf() returns null in v3.
+      while (
+        data != null &&
```
Comment (rangadi) [inline connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufCatalystDataConversionSuite.scala:136]: This should just be `data == null || data.asInstanceOf[Row].get(0) == defaultValue`?

Comment (rangadi) [reply connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufCatalystDataConversionSuite.scala:136]: I meant, we don't need the array check.

Comment (SandishKumarHN) [reply connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufCatalystDataConversionSuite.scala:136]: @rangadi 
1. `data.asInstanceOf[Row].get(0) == ByteString.empty().toByteArray`
2. `data.asInstanceOf[Row].get(0) == Array.emptyByteArray`
3. `data.asInstanceOf[Row].get(0) ==  ByteString.EMPTY.toByteArray`
4. `data.asInstanceOf[Row].get(0) ==   "".getBytes`
5. `data.asInstanceOf[Row].get(0).isInstanceOf[Array[Byte]] && data.asInstanceOf[Row].get(0).asInstanceOf[Array[Byte]].isEmpty`

Except for (5), none of them worked. I'm printing under the conditions listed above.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 3]
```diff
@@ -3330,7 +3330,7 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase {
   def descrioptorParseError(descFilePath: String, cause: Throwable): Throwable = {
     new AnalysisException(
       errorClass = "CANNOT_PARSE_PROTOBUF_DESCRIPTOR",
-      messageParameters = Map.empty("descFilePath" -> descFilePath),
+      messageParameters = Map("descFilePath" -> descFilePath),
```
Comment (SandishKumarHN) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:3333]: @MaxGekk fixing this one https://github.com/apache/spark/pull/38344#discussion_r1014644624 here

File: connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufSerdeSuite.scala [hunk 4]
```diff
@@ -177,6 +177,19 @@ class ProtobufSerdeSuite extends SharedSparkSession {
...
+  test("raise cannot parse protobuf descriptor error") {
+    // passing serde_suite.proto instead serde_suite.desc
+    val testFileDesc = testFile("serde_suite.proto").replace("file:/", "/")
+    val e = intercept[AnalysisException] {
+      ProtobufUtils.buildDescriptor(testFileDesc, "FieldMissingInSQLRoot")
+    }
+
+    checkError(
+      exception = e,
+      errorClass = "CANNOT_PARSE_PROTOBUF_DESCRIPTOR",
```
Comment (MaxGekk) [inline connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufSerdeSuite.scala:189]: @SandishKumarHN Thank you for the added test.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 5]
```diff
@@ -3344,7 +3344,7 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase {
   def failedParsingDescriptorError(descFilePath: String, cause: Throwable): Throwable = {
     new AnalysisException(
       errorClass = "CANNOT_CONSTRUCT_PROTOBUF_DESCRIPTOR",
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:3346]: Could you add a test for this error class.

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala (hunk 5 above)
Comment (SandishKumarHN) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:3346]: @MaxGekk added

File: connector/protobuf/src/test/resources/protobuf/basicmessage_noimports.desc [hunk 6]
```diff
@@ -0,0 +1,18 @@
+
```
Comment (rangadi) [inline connector/protobuf/src/test/resources/protobuf/basicmessage_noimports.desc:1]: Where is this used?
Did you mean to another test?

File: connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufSerdeSuite.scala [hunk 7]
```diff
@@ -177,6 +177,31 @@ class ProtobufSerdeSuite extends SharedSparkSession {
...
+      parameters = Map("descFilePath" -> testFileDesc))
+  }
+
+  test("raise cannot construct protobuf descriptor error") {
+    val testFileDesc = testFile("basicmessage_noimports.desc").replace("file:/", "/")
+    val e = intercept[AnalysisException] {
+      ProtobufUtils.parseFileDescriptorSet(testFileDesc)
```
[the remark this answers is not in the document]
Comment (SandishKumarHN) [reply connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufSerdeSuite.scala:196]: @rangadi  Yes, but the query would catch this error and throw a different error right?

File: connector/protobuf/src/test/resources/protobuf/basicmessage_noimports.desc (hunk 6 above)
Comment (SandishKumarHN) [reply connector/protobuf/src/test/resources/protobuf/basicmessage_noimports.desc:1]: @rangadi yes, it is used in the unit test above, how else can we raise the CANNOT_CONSTRUCT_PROTOBUF_DESCRIPTOR exception?

File: connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufSerdeSuite.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (rangadi) [reply connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufSerdeSuite.scala:196]: Users don't call this. Analysis exception thrown to the user if they try to use this.

File: connector/protobuf/src/test/resources/protobuf/basicmessage_noimports.desc (hunk 6 above)
Comment (rangadi) [reply connector/protobuf/src/test/resources/protobuf/basicmessage_noimports.desc:1]: Right. I didn't see the second test carefully.

File: connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufSerdeSuite.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (SandishKumarHN) [reply connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufSerdeSuite.scala]: @rangadi Check it out once again I've made a few minor changes.

File: connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufFunctionsSuite.scala [hunk 8]
```diff
@@ -677,4 +677,34 @@ class ProtobufFunctionsSuite extends QueryTest with SharedSparkSession with Seri
           === inputDf.select("durationMsg.duration").take(1).toSeq(0).get(0))
     }
   }
+
+  test("raise cannot construct protobuf descriptor error") {
+    val basicMessageDesc = ProtobufUtils.buildDescriptor(testFileDesc, "BasicMessage")
+
+    val basicMessage = DynamicMessage
```
Comment (rangadi) [inline connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufFunctionsSuite.scala:684]: Btw, you don't need the message. Could use empty byte array while initializing `df`.

**Review by rangadi — APPROVED** [review]

File: connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufFunctionsSuite.scala (hunk 8 above)
Comment (SandishKumarHN) [reply connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufFunctionsSuite.scala]: correct, fixed.

**Comment by rangadi** [comment]
@MaxGekk, please merge once this this looks good to you.

**Comment by MaxGekk** [comment]
@SandishKumarHN @rangadi I will review and merge it (if everything is ok, and test passed) tomorrow.

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @SandishKumarHN and @rangadi for review.

**Comment by SandishKumarHN** [comment]
@MaxGekk can also close the [JIRA](https://issues.apache.org/jira/browse/SPARK-41015)

**Pull request closed without merging**